In [63]:
import numpy as np
import pandas as pd
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
from sklearn.metrics import classification_report, confusion_matrix

In [64]:
Smarket = pd.read_csv("/Users/arpanganguli/Documents/Finance/ISLR/Datasets/Smarket.csv", index_col = 'SlNo')

In [65]:
Smarket.head()

Year   Lag1   Lag2   Lag3   Lag4   Lag5  Volume  Today Direction
SlNo                                                                  
1     2001  0.381 -0.192 -2.624 -1.055  5.010  1.1913  0.959        Up
2     2001  0.959  0.381 -0.192 -2.624 -1.055  1.2965  1.032        Up
3     2001  1.032  0.959  0.381 -0.192 -2.624  1.4112 -0.623      Down
4     2001 -0.623  1.032  0.959  0.381 -0.192  1.2760  0.614        Up
5     2001  0.614 -0.623  1.032  0.959  0.381  1.2057  0.213        Up

In [66]:
from sklearn.model_selection import train_test_split

In [67]:
X = np.array(Smarket[['Lag1', 'Lag2']])
y = np.array(Smarket['Direction'])
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2016, random_state=101)

In [68]:
ldafit = LDA()
ldafit.fit(X_train, y_train)

LinearDiscriminantAnalysis(n_components=None, priors=None, shrinkage=None,
              solver='svd', store_covariance=False, tol=0.0001)

In [69]:
ldafit.priors_

array([0.48496994, 0.51503006])

In [70]:
ldafit.means_

array([[ 0.06719008,  0.0221157 ],
       [-0.0297607 , -0.04227237]])

In [71]:
ldafit.coef_

array([[-0.07716148, -0.05114827]])

In [72]:
ldapred = ldafit.predict(X_test)

In [73]:
print(confusion_matrix(y_test, ldapred))

[[ 30  88]
 [ 23 111]]


In [74]:
print(classification_report(y_test, ldapred))

              precision    recall  f1-score   support

        Down       0.57      0.25      0.35       118
          Up       0.56      0.83      0.67       134

   micro avg       0.56      0.56      0.56       252
   macro avg       0.56      0.54      0.51       252
weighted avg       0.56      0.56      0.52       252



**The precision is intuitively the ability of the classifier not to label as positive a sample that is negative. Therefore,
this model is able to correctly classify 57% of the test data (which I believe is great given the vagaries of the stock 
market!!**